src/log.rs

```diff
@@ -226,3 +263,18 @@ pub(crate) fn debug_remove_log_function() {
         gst_debug_remove_log_function(Some(log_callback));
     }
 }
+
+#[inline]
+fn span_quark() -> &'static gstreamer::glib::Quark {
+    static ELEMENT_SPAN_QUARK: OnceCell<gstreamer::glib::Quark> = OnceCell::new();
+    ELEMENT_SPAN_QUARK.get_or_init(|| {
+        let gstr = gstreamer::glib::GStr::from_utf8_with_nul(b"tracing-gstreamer:span\0").unwrap();
+        gstreamer::glib::Quark::from_str(gstr)
+    })
+}
+
+pub(crate) fn attach_span<O: IsA<gstreamer::Object>>(object: &O, span: tracing::Span) {
+    unsafe {
+        object.set_qdata(*span_quark(), span);
```

Comment:
Hm, some work will need to happen before we can declare this sound, I worry. Some problematic cases I can think of by reading documenation on `set_qdata` and `qdata`:

1. Somebody directly overwrites the well-known key with arbitrary type of their own:
    * We should wrap `tracing::Span` inside a private newtype and append a `type_id()` to the `Quark` string to get at least a somewhat reasonable opportunistic protection from this data being overwritten.
2. Somebody runs this function in a loop in a separate thread on an object that's being inspected in `log_callback`
    * Do we need to take `object`’s lock in both places?

Reply:
> Somebody directly overwrites the well-known key with arbitrary type of their own:

I think what you added with the Quark name mangling is enough.

> Somebody runs this function in a loop in a separate thread on an object that's being inspected in `log_callback`

Good point.  Looking at the code `g_object_get_qdata()` and `g_object_set_qdata()`  are thread safe but that does not protect us from the returned data being destroyed as `g_object_get_qdata()`  does not duplicate it.

Taking the object lock would prevent that, but I'm afraid it may introduce dead locks if logs are produced while this lock is hold by the caller (the mutex is not recursive).

@sdroege : does that make sense or am I missing something here? Or do you have any smart idea? :)

One option would be to stop using `qdata` and goes back to the `HashMap<Weak<GstObject>, Span>` plan which could be protected by its own mutex. I know you'd rather avoid it but that may be the safest option.

Reply:
@nagisa I'd like to move forward with this. What would you suggest?

Reply:
I think the best way forward for the time being is to make `attach_span` `unsafe` and add some documentation as to why it is the way it is. I think avoiding data races is largely trivial in most sensible uses of this function as I imagine this function is going to be called to attach a span before the pipeline or elements are running (e.g. on a program startup.)

So basically

```
# Safety

This function is racy and must not be called when an element or any of its children may want to emit a log message. It is safe to use this function when an element is in a null (add others if there are any) state.
```

Reply:
Either way I’ll merge this tonight and make the change myself if you don't get around to making the change yourself by then.

Reply:
Sounds good to me.

I made the change and squashed commits. I also rebased it on top of #19 so you may want to merge this one first.